crates/sequencer/src/sequencer.rs

```diff
@@ -564,26 +420,63 @@ where
-                    }
+        // Initialize our knowledge of the state of the DA-layer
+        let (mut prev_l1_height, mut last_finalized_block, mut l1_fee_rate) =
+            match get_da_block_data(self.da_service.clone(), self.ledger_db.clone()).await {
+                Ok(l1_data) => l1_data,
+                Err(e) => {
+                    error!("{}", e);
+                    return Err(e);
                 }
-            }
+            };
+
+        // Based on the above initialization, decide whether we should submit a commitment to DA.
+        if let Err(e) = self
+            .submit_commitment(last_finalized_block.header().height(), prev_l1_height)
+            .await
+        {
+            error!("Sequencer error: {}", e);
         }
-        // If sequencer is in production mode, it will build a block every 2 seconds
-        else {
-            loop {
-                sleep(Duration::from_secs(2)).await;
-                if let Err(e) = self.build_block().await {
-                    error!("Sequencer error: {}", e);
+
+        // Setup required workers to update our knowledge of the DA layer every 2 seconds.
+        let (da_tx, mut da_rx) = mpsc::channel(1);
+        let da_monitor = da_block_monitor(self.da_service.clone(), self.ledger_db.clone(), da_tx);
+        tokio::pin!(da_monitor);
+        let mut interval = tokio::time::interval(Duration::from_secs(2));
+
+        loop {
+            tokio::select! {
+                // Run the DA monitor worker
+                _ = &mut da_monitor => {},
+                // Receive updates from DA layer worker.
+                l1_data = da_rx.recv() => {
+                    if let Some(l1_data) = l1_data {
+                        (prev_l1_height, last_finalized_block, l1_fee_rate) = l1_data;
+                        if let Err(e) = self.submit_commitment(last_finalized_block.header().height(), prev_l1_height).await {
+                            error!("Sequencer error: {}", e);
+                        }
+                    }
+                },
+                // If sequencer is in test mode, it will build a block every time it receives a message
+                // The RPC from which the sender can be called is only registered for test mode. This means
+                // that evey though we check the receiver here, it'll never be "ready" to be consumed unless in test mode.
+                _ = self.l2_force_block_rx.next() => {
+                    if self.config.test_mode {
+                        if let Err(e) = self.produce_l2_block(last_finalized_block.clone(), l1_fee_rate).await {
+                            error!("Sequencer error: {}", e);
+                        }
+                    }
+                },
+                // If sequencer is in production mode, it will build a block every 2 seconds
+                _ = interval.tick() => {
+                    if let Err(e) = self.produce_l2_block(last_finalized_block.clone(), l1_fee_rate).await {
```

Comment:
I actually wanted to focus more on making the time between end of block n-1 (publishing) and end of block n 2 seconds. (or their starts)

(produce_l2_block takes some time you know)

Reply:
Addressed.